File: python_port/rag/impl/rule_intent_detector.py
from typing import List, Dict, Optional
from ..interfaces import IntentDetector, Intent

class RuleIntentDetector(IntentDetector):
    def __init__(self, rules: Dict[str, List[str]]):
        self.rules = rules

    def detect(self, question: str) -> Intent:
        lower_question = question.lower()

        for intent in Intent:
            if intent == Intent.Unknown:
                continue
            
            keywords = self.rules.get(intent.value)
            if keywords:
                for keyword in keywords:
                    if keyword.lower() in lower_question:
                        return intent
        
        return Intent.Unknown

File: Iteration1CSE3063Proje/src/main/java/com/cse3063/rag/impl/RuleIntentDetector.java
package com.cse3063.rag.impl;

import com.cse3063.rag.interfaces.IntentDetector;
import java.util.List;
import java.util.Map;

public class RuleIntentDetector implements IntentDetector {
    private final Map<String, List<String>> rules;

    public RuleIntentDetector(Map<String, List<String>> rules) {
        this.rules = rules;
    }

    @Override
    public Intent detect(String question) {
        String lowerQuestion = question.toLowerCase();

        // Priority can be configured, iterating map for now
        for (Intent intent : Intent.values()) {
            if (intent == Intent.Unknown)
                continue;

            List<String> keywords = rules.get(intent.name());
            if (keywords != null) {
                for (String keyword : keywords) {
                    if (lowerQuestion.contains(keyword.toLowerCase())) {
                        return intent;
                    }
                }
            }
        }

        return Intent.Unknown;
    }
}
